File: SWEA/D2/1954. 달팽이 숫자/달팽이 숫자.py
T = int(input())

dx = [0,1,0,-1]
dy = [1,0,-1,0]

def DFS(matrix, direction, x, y, num):
    if x > len(matrix)-1 or y > len(matrix[0])-1 or matrix[x][y] != 0:
        return
    else:
        matrix[x][y] += num
        if 0 <= x+dx[direction] < len(matrix) and 0 <= y+dy[direction] < len(matrix[0]) and matrix[x+dx[direction]][y+dy[direction]] == 0:
            pass
        else:
            direction += 1
            direction %= 4
        
        DFS(matrix, direction, x+dx[direction], y+dy[direction], num+1)

# 여러개의 테스트 케이스가 주어지므로, 각각을 처리합니다.
for test_case in range(1, T + 1):
    # ///////////////////////////////////////////////////////////////////////////////////
    N = int(input())
    matrix = [[0 for _ in range(N)] for _ in range(N)]
    DFS(matrix, 0, 0, 0, 1)
    
    print(f"#{test_case}")

    for i in range(N):
        for j in range(N):
            print(matrix[i][j], end=' ')
            if j==(N-1):
                print()

File: SWEA/D2/1954. 달팽이 숫자/달팽이 숫자.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Solution {
	static int[] dx = {0,1,0,-1};
	static int[] dy = {1,0,-1,0};
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int T = Integer.parseInt(br.readLine());
		for(int test_case = 1; test_case <= T; test_case++) {
			int n = Integer.parseInt(br.readLine());
			int [][] matrix = new int[n][n];
			for(int i = 0; i < n; i++) {
				for(int j = 0; j < n; j++) {
					matrix[i][j] = 0;
				}
			}
			
			int direction = 0;
			int num = 1;
			int x = 0;
			int y = 0;
			
			for(int i = 0; i < n*n; i++) {
				matrix[x][y] = num;
				num ++;
				
				if(x+dx[direction] < 0 || x+dx[direction]>=n || y+dy[direction] < 0 || y+dy[direction]>=n) {
					direction += 1;
					direction %= 4;
				}else if(matrix[x+dx[direction]][y+dy[direction]] != 0) {
					direction += 1;
					direction %= 4;
				}
				
				x += dx[direction];
				y += dy[direction];
				
			}
            
            System.out.println("#"+test_case);
			
			for(int i = 0; i < n; i++) {
				for(int j = 0; j < n; j++) {
					System.out.print(matrix[i][j]+" ");
				}
				System.out.println();
			}
		}
	}
}
